Transition to `CGEvent.tapCreate` for hotkey bindings (#1012)

NMAC427 opened the pull request:
Replace the Carbon based HotKey library with `CGEvent.tapCreate` as layed out in #1012.
As a consequence, we get direct access to all key-down events, and can choose to handle them based on the current modifier state and key code. Therefore, in addition to #1012, this change also addresses all of the following issues / feature requests:

- #28
- #1011
- #1472

The primary user facing changes are as follows:
- Added new modifiers `lcmd`, `rcmd`, `lalt`, `ralt`, `lctrl`, `rctrl`, `lshift`, `rshift`, and `fn`.
- Add `key-mapping.match-key-event-by` to choose between matching events based on key codes or key symbols.

## PR checklist

- [x] Explain your changes in the relevant commit messages rather than in the PR description. The PR description must not contain more information than the commit messages (except for images and other media).
- [x] Each commit must explain what/why/how and motivation in its description. https://cbea.ms/git-commit/
- [x] Don't forget to link the appropriate issues/discussions in commit messages (if applicable).
- [x] Each commit must be an atomic change (a PR may contain several commits). Don't introduce new functional changes together with refactorings in the same commit.
- [x] `./run-tests.sh` exits with non-zero exit code.
- [x] Avoid merge commits, always rebase and force push.

Failure to follow the checklist with no apparent reasons will result in silent PR rejection.

nikitabobko commented:
Previous PR: https://github.com/nikitabobko/AeroSpace/pull/1691

---

1. Please avoid merge commits. Use only rebase and force push.
2. The commits must be atomic. The first commit must be the migration from Carbon API to `CGEvent.tapCreate` and nothing else. Right now, you do the migration, change semantics, add new features, and do breaking changes all together in a single commit.
3. Each commit must have its respective git description.
4. Please also take a look at https://github.com/nikitabobko/AeroSpace/blob/main/.github/pull_request_template.md

Answering questions from your [previous PR](https://github.com/nikitabobko/AeroSpace/pull/1691)

> However, before I get started, I would like to get your opinion: Do you foresee any reason why we should keep around any of the keymap functionality?

Yes, we should keep the keymap functionality. Languages and keyboard layouts are wild beasts full of corner cases and exceptions. I am sure that there are people who prefer the current behavior of matching by keycodes (I am kinda such person as I type in qwerty and cyrillic).

The current default of matching by key codes should stay at least for compatibility reasons.

> This usually isn't a big problem, as `key-mapping.key-notation-to-key-code` can be used to adjust the layout. However, when working with a custom keyboard that is remapped on the firmware level (i.e. anything running ZMK), in conjuncion with a keyboard that is remapped in software (i.e. builtin MacBook keyboard), this results in significant issues.

I find this use case unconvincing. You have to switch input sources when you connect your external keyboard anyway. If it's the only use case that we have on the table, I am hesitant to optimize for it. If somebody has other use cases, they are welcome to describe them.

I like the current algorithm of matching by key codes + manual remapping via `key-mapping.key-notation-to-key-code`, because it covers 99% of use cases, and it offloads the responsibility of wild west layouts onto the user.

I imagine a config option that would allow to match by key symbols instead of key codes:
```toml
[key-mapping]
  # Possible values: (key-codes|key-symbols)
  match-key-event-by = 'key-codes' # The current default
```

but I am still not convinced in its necessity.

NMAC427 committed:
Add `key-mapping.match-key-event-by` with `key-code` (default) and new `key-symbol` option (fixes #1472).

Motivation:
AeroSpace currently interprets all key bindings using a QWERTY-based key-code model. While `key-mapping.key-notation-to-key-code` can compensate for software layout changes, this approach breaks down when mixing keyboards that are remapped at different layers.
In particular, when using a hardware-remapped keyboard (e.g. ZMK firmware) alongside a software-remapped keyboard (e.g. the built-in MacBook keyboard), identical logical key presses can produce different key codes. As a result, the same shortcut may work on one keyboard but silently fail on the other. `key-mapping.match-key-event-by` allows key bindings to be matched either by key code (current behavior) or by key symbol. This ensures consistent shortcut behavior across mixed hardware/software keyboard setups, while also providing a simpler alternative to `key-mapping.preset`.

NMAC427 committed:
Add `key-mapping.match-key-event-by` with `key-code` (default) and new `key-symbol` option (fixes #1472).

Motivation:
AeroSpace currently interprets all key bindings using a QWERTY-based key-code model. While `key-mapping.key-notation-to-key-code` can compensate for software layout changes, this approach breaks down when mixing keyboards that are remapped at different layers.
In particular, when using a hardware-remapped keyboard (e.g. ZMK firmware) alongside a software-remapped keyboard (e.g. the built-in MacBook keyboard), identical logical key presses can produce different key codes. As a result, the same shortcut may work on one keyboard but silently fail on the other. `key-mapping.match-key-event-by` allows key bindings to be matched either by key code (current behavior) or by key symbol. This ensures consistent shortcut behavior across mixed hardware/software keyboard setups, while also providing a simpler alternative to `key-mapping.preset`.

NMAC427 committed:
Implement differentiating between left and right modifier keys (fixes #28), and support fn as modifier (fixes #1011).

Motivation: Many international keyboard layouts rely on the right Alt key (AltGr, https://en.wikipedia.org/wiki/AltGr_key) for an additional symbol layer. Aerospace not differentiating between Alt and AltGr prevents users from using alt for hot

NMAC427 committed:
Implement differentiating between left and right modifier keys (fixes #28), and support fn as modifier (fixes #1011).

Motivation: Many international keyboard layouts rely on the right Alt key (AltGr, https://en.wikipedia.org/wiki/AltGr_key) for an additional symbol layer. Aerospace not differentiating between Alt and AltGr prevents users from using alt for hot

NMAC427 committed:
Replace HotKey (Carbon API) with newer `CGEvent.tapCreate` API (fixes #1012).

Motivation: The Carbon API has been deprecated, and the `CGEvent.tapCreate` API will eventually allow us to differentiate between left and right modifier keys.

NMAC427 committed:
Replace HotKey (Carbon API) with newer `CGEvent.tapCreate` API (fixes #1012).

Motivation: The Carbon API has been deprecated, and the `CGEvent.tapCreate` API will eventually allow us to differentiate between left and right modifier keys.

kfiz commented:
@nikitabobko I've been looking to get builds of this as suggested  here https://github.com/nikitabobko/AeroSpace/blob/main/dev-docs/development.md at https://github.com/nikitabobko/AeroSpace/actions/runs/20724817788 . Are there supposed to be build artifacts somewhere that I could download for testing?

exlee commented:
I'm yet another person that would like to see this move forward.

It covers gripes I have about aerospace (and these seem to be the only ones):

- AltGr (right now I have "passthrough mode" to type, cumbersome)
- Small shortcut surface (which would be solved by treating LCmd and RCmd differently as well as possibility of Fn key use)

Looking forward to be able to specify LAlt-Arrows to map to something other than RAlt-Arrows.

nulam commented:
> > RE: I was able to build it from your branch @NMAC427. I can confirm that it works flawlessly. Thanks! Can't wait for this to get merged.
> 
> @GaetanLepage Care to share how you built it. ;) I was trying to build the .app with nix but didn't succeed in doing so, the cli binary built successfully though. I'm hesitant to pull in all external deps needed by the app by just running the provided build scripts as I ideally would like nix to manage all my deps.

Not being able to differentiate left and right modifiers is breaking to my keyboard layout with diacritics, hence I built it myself, can confirm it works just fine. 

I got it running with Claude and I let it write up a [tutorial](https://pastebin.com/XvBikexc) afterwards, no guarantees or anything, but it works on my end.

Please note that this tutorial uses the homebrew folder, which might conflict when upgrading via brew. Either tweak the folders or uninstall the default aerospace first.

Edit: disclaimer - i have found this build of aerospace sometimes puts constant 99% load on my CPU, i haven't yet identified the cause, but it severely harms my battery life :/

LeszekBlazewski commented:
@nikitabobko Looking forward for this feature as well!
